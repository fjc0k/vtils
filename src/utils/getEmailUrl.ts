import { createUrlQueryString } from './createUrlQueryString'

export interface GetEmailUrlOptions {
  /** 收件人 */
  to?: string | string[]
  /** 抄送人 */
  cc?: string | string[]
  /** 密送人 */
  bcc?: string | string[]
  /** 主题 */
  subject?: string
  /** 内容 */
  body?: string
}

/**
 * 获取邮件链接。
 *
 * @param options 选项
 */
export function getEmailUrl(options: GetEmailUrlOptions): string {
  const { to, cc, bcc, subject, body } = options
  let url = `mailto:${Array.isArray(to) ? to.join(',') : to}`
  if (cc || bcc || subject || body) {
    url += `?${createUrlQueryString({
      cc: cc && (Array.isArray(cc) ? cc.join(',') : cc),
      bcc: bcc && (Array.isArray(bcc) ? bcc.join(',') : bcc),
      subject: subject,
      body: body,
    })}`
  }
  return url
}
