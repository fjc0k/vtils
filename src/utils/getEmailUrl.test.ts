import { getEmailUrl } from './getEmailUrl.ts'

describe('getEmailUrl', () => {
  test('ok', () => {
    expect(
      getEmailUrl({
        to: 'hello@gmail.com',
      }),
    ).toBe('mailto:hello@gmail.com')
    expect(
      getEmailUrl({
        to: ['hello@gmail.com', 'hello2@gmail.com'],
      }),
    ).toBe('mailto:hello@gmail.com,hello2@gmail.com')
    expect(
      getEmailUrl({
        to: 'hello@gmail.com',
        cc: 'hello2@gmail.com',
        bcc: 'hello3@gmail.com',
        subject: 'hello',
        body: '你好',
      }),
    ).toMatchSnapshot()
    expect(
      getEmailUrl({
        to: 'hello@gmail.com',
        cc: ['hello2@gmail.com', 'hello22@gmail.com'],
        bcc: ['hello3@gmail.com', 'hello333@gmail.com'],
        subject: 'hello',
        body: '你好',
      }),
    ).toMatchSnapshot()
  })
})
