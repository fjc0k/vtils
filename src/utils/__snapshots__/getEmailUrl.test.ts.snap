// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEmailUrl ok 1`] = `"mailto:hello@gmail.com?cc=hello2%40gmail.com&bcc=hello3%40gmail.com&subject=hello&body=%E4%BD%A0%E5%A5%BD"`;

exports[`getEmailUrl ok 2`] = `"mailto:hello@gmail.com?cc=hello2%40gmail.com%2Chello22%40gmail.com&bcc=hello3%40gmail.com%2Chello333%40gmail.com&subject=hello&body=%E4%BD%A0%E5%A5%BD"`;
